packages/json-rpc/ws-provider/src/follow-enhancer.ts

```diff
@@ -0,0 +1,86 @@
+import { JsonRpcProvider } from "@polkadot-api/json-rpc-provider"
+import { WsJsonRpcProvider } from "./types"
+
+const methods: Record<string, "follow" | "unfollow"> = {}
+;["v1", "unstable"].forEach((version) => {
+  methods[`chainHead_${version}_follow`] = "follow"
+  methods[`chainHead_${version}_unfollow`] = "unfollow"
+})
+
+export const followEnhancer: (input: WsJsonRpcProvider) => WsJsonRpcProvider = (
+  base,
+) => {
+  const { getStatus } = base
+  const prematureStops = new Set<string>()
+  const preOpId = new Set<string>()
+  const onGoing = new Set<string>()
+
+  const clear = () => {
+    prematureStops.clear()
+    preOpId.clear()
+    onGoing.clear()
+  }
+  const enhancedSwitch: typeof base.switch = (...args) => {
+    clear()
+    base.switch(...args)
+  }
```

Comment:
The enhanced switch will only run when being called from outside or within this enhancer, but it won't be called when the `wsProvider` calls `onHalt` to start a new connection (probably jumping to the next in the ws list)

Is this important?

Reply:
good catch! fixed! Thanks!